types/yargs/index.d.ts

```diff
@@ -835,25 +835,51 @@ declare namespace yargs {
 
     // prettier-ignore
     type InferredOptionType<O extends Options | PositionalOptions> =
+        // Handle special cases first
+        O extends (
+            | { coerce: (arg: any) => infer T }
+        ) ? IsRequiredOrHasDefault<O> extends true ? T : T | undefined :
+        O extends (
+            | { type: "count"; default: infer D }
+            | { count: true; default: infer D }
+        ) ? number | Exclude<D, undefined> :
+        O extends (
+            | { type: "count" }
+            | { count: true }
+        ) ? number :
+        // Try to infer type with InferredOptionTypePrimitive
+        IsUnknown<InferredOptionTypePrimitive<O>> extends false ? InferredOptionTypePrimitive<O> :
+        // Use the type of `default` as the last resort
+        O extends (
+            | { default: infer D }
+        ) ? Exclude<D, undefined> : unknown;
+
+    // prettier-ignore
+    type IsRequiredOrHasDefault<O extends Options | PositionalOptions> =
         O extends (
             | { required: string | true }
             | { require: string | true }
             | { demand: string | true }
             | { demandOption: string | true }
-        ) ?
-        Exclude<InferredOptionTypeInner<O>, undefined> :
-        InferredOptionTypeInner<O>;
+            | { default: {} }
+        ) ? true : false;
 
+    type IsAny<T> = 0 extends (1 & T) ? true : false;
     // prettier-ignore
-    type InferredOptionTypeInner<O extends Options | PositionalOptions> =
-        O extends { default: any, coerce: (arg: any) => infer T } ? T :
-        O extends { default: infer D } ? D :
-        O extends { type: "count" } ? number :
-        O extends { count: true } ? number :
-        RequiredOptionType<O> | undefined;
+    type IsUnknown<T> =
+        IsAny<T> extends true ? false :
+        unknown extends T ? true :
+        false;
+
+    // prettier-ignore
+    type InferredOptionTypePrimitive<O extends Options | PositionalOptions> =
+        O extends { default: infer D } ? InferredOptionTypeInner<O> | D :
+        IsRequiredOrHasDefault<O> extends true ? InferredOptionTypeInner<O> :
+        InferredOptionTypeInner<O> | undefined;
 
     // prettier-ignore
-    type RequiredOptionType<O extends Options | PositionalOptions> =
```

Comment:
@ciffelia `index.d.mts` tries to reexport this type and produces errors for consumers with `node16` module resolution now:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/yargs/index.d.mts#L32

Reply:
@ciffelia This is a **blocking** issue for any ESM projects using this type declaration - tsc is unable to compile. Can you please submit a fix?